File: sample/OpenCvSharp.Samples/Samples/ConvexityDefect.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using OpenCvSharp;
using OpenCvSharp.Blob;

namespace OpenCvSharp.Test
{
    /// <summary>
    /// cvConvextyDefects sample
    /// </summary>
    class ConvexityDefect
    {
        public ConvexityDefect()
        {
            using (IplImage imgSrc = new IplImage(Const.ImageHand, LoadMode.Color))
            using (IplImage imgHSV = new IplImage(imgSrc.Size, BitDepth.U8, 3))
            using (IplImage imgH = new IplImage(imgSrc.Size, BitDepth.U8, 1))
            using (IplImage imgS = new IplImage(imgSrc.Size, BitDepth.U8, 1))
            using (IplImage imgV = new IplImage(imgSrc.Size, BitDepth.U8, 1))
            using (IplImage imgBackProjection = new IplImage(imgSrc.Size, BitDepth.U8, 1))     
            using (IplImage imgFlesh = new IplImage(imgSrc.Size, BitDepth.U8, 1))
            using (IplImage imgHull = new IplImage(imgSrc.Size, BitDepth.U8, 1))
            using (IplImage imgDefect = new IplImage(imgSrc.Size, BitDepth.U8, 3))
            using (IplImage imgContour = new IplImage(imgSrc.Size, BitDepth.U8, 3))
            using (CvMemStorage storage = new CvMemStorage())
            {
                // RGB -> HSV
                Cv.CvtColor(imgSrc, imgHSV, ColorConversion.BgrToHsv);
                Cv.CvtPixToPlane(imgHSV, imgH, imgS, imgV, null);
                IplImage[] hsvPlanes = { imgH, imgS, imgV };

                // 肌色領域を求める
                RetrieveFleshRegion(imgSrc, hsvPlanes, imgBackProjection);
                // 最大の面積の領域を残す
                FilterByMaximalBlob(imgBackProjection, imgFlesh);
                Interpolate(imgFlesh);

                // 輪郭を求める
                CvSeq<CvPoint> contours = FindContours(imgFlesh, storage);
                if (contours != null)
                {
                    Cv.DrawContours(imgContour, contours, CvColor.Red, CvColor.Green, 0, 3, LineType.AntiAlias);

                    // 凸包を求める
                    int[] hull;
                    Cv.ConvexHull2(contours, out hull, ConvexHullOrientation.Clockwise);
                    Cv.Copy(imgFlesh, imgHull);
                    DrawConvexHull(contours, hull, imgHull);

                    // 凹状欠損を求める
                    Cv.Copy(imgContour, imgDefect);
                    CvSeq<CvConvexityDefect> defect = Cv.ConvexityDefects(contours, hull);
                    DrawDefects(imgDefect, defect);
                }

                using (new CvWindow("src", imgSrc))
                using (new CvWindow("back projection", imgBackProjection))
                using (new CvWindow("hull", imgHull))
                using (new CvWindow("defect", imgDefect))
                {
                    Cv.WaitKey();
                }
            }
        }

        /// <summary>
        /// バックプロジェクションにより肌色領域を求める
        /// </summary>
        /// <param name="imgSrc"></param>
        /// <param name="hsvPlanes"></param>
        /// <param name="imgRender"></param>
        private void RetrieveFleshRegion(IplImage imgSrc, IplImage[] hsvPlanes, IplImage imgDst)
        {
            int[] histSize = new int[] { 30, 32 };
            float[] hRanges = { 0.0f, 20f };
            float[] sRanges = { 50f, 255f };
            float[][] ranges = { hRanges, sRanges };

            imgDst.Zero();
            using (CvHistogram hist = new CvHistogram(histSize, HistogramFormat.Array, ranges, true))
            {
                hist.Calc(hsvPlanes, false, null);
                float minValue, maxValue;
                hist.GetMinMaxValue(out minValue, out maxValue);
                hist.Normalize(imgSrc.Width * imgSrc.Height * 255 / maxValue);
                hist.CalcBackProject(hsvPlanes, imgDst);
            }
        }
        /// <summary>
        /// ラベリングにより最大の面積の領域を残す
        /// </summary>
        /// <param name="imgSrc"></param>
        /// <param name="imgRender"></param>
        private void FilterByMaximalBlob(IplImage imgSrc, IplImage imgDst)
        {
            using (CvBlobs blobs = new CvBlobs())
            using (IplImage imgLabelData = new IplImage(imgSrc.Size, CvBlobLib.DepthLabel, 1))
            {
                imgDst.Zero();
                blobs.Label(imgSrc, imgLabelData);
                CvBlob max = blobs[blobs.GreaterBlob()];
                if (max == null)
                {
                    return;
                }
                blobs.FilterByArea(max.Area, max.Area);
                blobs.FilterLabels(imgLabelData, imgDst);
            }
        }
        /// <summary>
        /// 欠損領域を補完する
        /// </summary>
        /// <param name="img"></param>
        private void Interpolate(IplImage img)
        {
            Cv.Dilate(img, img, null, 2);
            Cv.Erode(img, img, null, 2);
        }
        /// <summary>
        /// 輪郭を得る
        /// </summary>
        /// <param name="img"></param>
        /// <param name="storage"></param>
        /// <returns></returns>
        private CvSeq<CvPoint> FindContours(IplImage img, CvMemStorage storage)
        {
            // 輪郭抽出
            CvSeq<CvPoint> contours;
            using (IplImage imgClone = img.Clone())
            {
                Cv.FindContours(imgClone, storage, out contours);
                if (contours == null)
                {
                    return null;
                }
                contours = Cv.ApproxPoly(contours, CvContour.SizeOf, storage, ApproxPolyMethod.DP, 3, true);
            }
            // 一番長そうな輪郭のみを得る
            CvSeq<CvPoint> max = contours;
            for (CvSeq<CvPoint> c = contours; c != null; c = c.HNext)
            {
                if (max.Total < c.Total)
                {
                    max = c;
                }
            }
            return max;
        }

        /// <summary>
        /// ConvexHullの描画
        /// </summary>
        /// <param name="contours"></param>
        /// <param name="hull"></param>
        /// <param name="img"></param>
        private void DrawConvexHull(CvSeq<CvPoint> contours, int[] hull, IplImage img)
        {
            CvPoint pt0 = contours[hull.Last()].Value;
            foreach (int idx in hull)
            {
                CvPoint pt = contours[idx].Value;
                Cv.Line(img, pt0, pt, new CvColor(255, 255, 255));
                pt0 = pt;
            }
        }
        /// <summary>
        /// ConvexityDefectsの描画
        /// </summary>
        /// <param name="img"></param>
        /// <param name="defect"></param>
        private void DrawDefects(IplImage img, CvSeq<CvConvexityDefect> defect)
        {
            int count = 0;
            foreach (CvConvexityDefect item in defect)
            {
                CvPoint p1 = item.Start, p2 = item.End;
                double dist = GetDistance(p1, p2);
                CvPoint2D64f mid = GetMidpoint(p1, p2);
                img.DrawLine(p1, p2, CvColor.White, 3);
                img.DrawCircle(item.DepthPoint, 10, CvColor.Green, -1);
                img.DrawLine(mid, item.DepthPoint, CvColor.White, 1);   
                Console.WriteLine("No:{0} Depth:{1} Dist:{2}", count, item.Depth, dist);
                count++;
            }
        }
        /// <summary>
        /// 2点間の距離を得る
        /// </summary>
        /// <param name="p1"></param>
        /// <param name="p2"></param>
        /// <returns></returns>
        private double GetDistance(CvPoint p1, CvPoint p2)
        {
            return Math.Sqrt(Math.Pow(p1.X - p2.X, 2) + Math.Pow(p1.Y - p2.Y, 2));
        }
        /// <summary>
        /// 2点の中点を得る
        /// </summary>
        /// <param name="p1"></param>
        /// <param name="p2"></param>
        /// <returns></returns>
        private CvPoint2D64f GetMidpoint(CvPoint p1, CvPoint p2)
        {
            return new CvPoint2D64f
            {
                X = (p1.X + p2.X) / 2.0,
                Y = (p1.Y + p2.Y) / 2.0
            };
        }
    }
}

File: sample/CStyleSamplesVB/Samples/ConvexityDefect.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Runtime.InteropServices
Imports System.Text
Imports OpenCvSharp
Imports OpenCvSharp.Blob

' Namespace OpenCvSharpSamplesVB
Imports SampleBase

''' <summary>
''' cvConvextyDefects sample
''' </summary>
Friend Module ConvexityDefect
    Public Sub Start()
        Using imgSrc As New IplImage(FilePath.Image.Hand, LoadMode.Color), _
             imgHSV As New IplImage(imgSrc.Size, BitDepth.U8, 3), _
         imgH As New IplImage(imgSrc.Size, BitDepth.U8, 1), _
             imgS As New IplImage(imgSrc.Size, BitDepth.U8, 1), _
             imgV As New IplImage(imgSrc.Size, BitDepth.U8, 1), _
            imgBackProjection As New IplImage(imgSrc.Size, BitDepth.U8, 1), _
            imgFlesh As New IplImage(imgSrc.Size, BitDepth.U8, 1), _
            imgHull As New IplImage(imgSrc.Size, BitDepth.U8, 1), _
            imgDefect As New IplImage(imgSrc.Size, BitDepth.U8, 3), _
            imgContour As New IplImage(imgSrc.Size, BitDepth.U8, 3), _
            storage As New CvMemStorage()
            ' RGB -> HSV
            Cv.CvtColor(imgSrc, imgHSV, ColorConversion.BgrToHsv)
            Cv.CvtPixToPlane(imgHSV, imgH, imgS, imgV, Nothing)
            Dim hsvPlanes() As IplImage = {imgH, imgS, imgV}

            RetrieveFleshRegion(imgSrc, hsvPlanes, imgBackProjection)
            ' Get max area blob
            FilterByMaximalBlob(imgBackProjection, imgFlesh)
            Interpolate(imgFlesh)

            ' find contours
            Dim contours As CvSeq(Of CvPoint) = FindContours(imgFlesh, storage)
            If contours IsNot Nothing Then
                Cv.DrawContours(imgContour, contours, CvColor.Red, CvColor.Green, 0, 3, LineType.AntiAlias)

                ' ConvexHull
                Dim hull() As Integer
                Cv.ConvexHull2(contours, hull, ConvexHullOrientation.Clockwise)
                Cv.Copy(imgFlesh, imgHull)
                DrawConvexHull(contours, hull, imgHull)

                ' ConvexityDefects
                Cv.Copy(imgContour, imgDefect)
                Dim defect As CvSeq(Of CvConvexityDefect) = Cv.ConvexityDefects(contours, hull)
                DrawDefects(imgDefect, defect)
            End If

            Using TempCvWindow As CvWindow = New CvWindow("src", imgSrc), _
                     TempCvWindowBack As CvWindow = New CvWindow("back projection", imgBackProjection), _
                 TempCvWindowHull As CvWindow = New CvWindow("hull", imgHull), _
                 TempCvWindowDefect As CvWindow = New CvWindow("defect", imgDefect)
                Cv.WaitKey()
            End Using

        End Using
    End Sub

    ''' <summary>
    ''' バックプロジェクションにより肌色領域を求める
    ''' </summary>
    ''' <param name="imgSrc"></param>
    ''' <param name="hsvPlanes"></param>
    ''' <param name="imgRender"></param>
    Private Sub RetrieveFleshRegion(ByVal imgSrc As IplImage, ByVal hsvPlanes() As IplImage, ByVal imgDst As IplImage)
        Dim histSize() As Integer = {30, 32}
        Dim hRanges() As Single = {0.0F, 20.0F}
        Dim sRanges() As Single = {50.0F, 255.0F}
        Dim ranges()() As Single = {hRanges, sRanges}

        imgDst.Zero()
        Using hist As New CvHistogram(histSize, HistogramFormat.Array, ranges, True)
            hist.Calc(hsvPlanes, False, Nothing)
            Dim minValue, maxValue As Single
            hist.GetMinMaxValue(minValue, maxValue)
            hist.Normalize(imgSrc.Width * imgSrc.Height * 255 / maxValue)
            hist.CalcBackProject(hsvPlanes, imgDst)
        End Using
    End Sub

    ''' <summary>
    ''' ラベリングにより最大の面積の領域を残す
    ''' </summary>
    ''' <param name="imgSrc"></param>
    ''' <param name="imgRender"></param>
    Private Sub FilterByMaximalBlob(ByVal imgSrc As IplImage, ByVal imgDst As IplImage)
        Dim blobs As New CvBlobs()

        imgDst.Zero()
        blobs.Label(imgSrc)
        Dim max As CvBlob = blobs.GreaterBlob()
        If max Is Nothing Then
            Return
        End If
        blobs.FilterByArea(max.Area, max.Area)
        blobs.FilterLabels(imgDst)
    End Sub

    ''' <summary>
    ''' 欠損領域を補完する
    ''' </summary>
    ''' <param name="img"></param>
    Private Sub Interpolate(ByVal img As IplImage)
        Cv.Dilate(img, img, Nothing, 2)
        Cv.Erode(img, img, Nothing, 2)
    End Sub
    ''' <summary>
    ''' 輪郭を得る
    ''' </summary>
    ''' <param name="img"></param>
    ''' <param name="storage"></param>
    ''' <returns></returns>
    Private Function FindContours(ByVal img As IplImage, ByVal storage As CvMemStorage) As CvSeq(Of CvPoint)
        ' 輪郭抽出
        Dim contours As CvSeq(Of CvPoint)
        Using imgClone As IplImage = img.Clone()
            Cv.FindContours(imgClone, storage, contours)
            If contours Is Nothing Then
                Return Nothing
            End If
            contours = Cv.ApproxPoly(contours, CvContour.SizeOf, storage, ApproxPolyMethod.DP, 3, True)
        End Using
        ' 一番長そうな輪郭のみを得る
        Dim max As CvSeq(Of CvPoint) = contours
        Dim c As CvSeq(Of CvPoint) = contours
        Do While c IsNot Nothing
            If max.Total < c.Total Then
                max = c
            End If
            c = c.HNext
        Loop
        Return max
    End Function

    ''' <summary>
    ''' ConvexHullの描画
    ''' </summary>
    ''' <param name="contours"></param>
    ''' <param name="hull"></param>
    ''' <param name="img"></param>
    Private Sub DrawConvexHull(ByVal contours As CvSeq(Of CvPoint), ByVal hull() As Integer, ByVal img As IplImage)
        Dim pt0 As CvPoint = contours(hull.Last()).Value
        For Each idx As Integer In hull
            Dim pt As CvPoint = contours(idx).Value
            Cv.Line(img, pt0, pt, New CvColor(255, 255, 255))
            pt0 = pt
        Next idx
    End Sub
    ''' <summary>
    ''' ConvexityDefectsの描画
    ''' </summary>
    ''' <param name="img"></param>
    ''' <param name="defect"></param>
    Private Sub DrawDefects(ByVal img As IplImage, ByVal defect As CvSeq(Of CvConvexityDefect))
        Dim count As Integer = 0
        For Each item As CvConvexityDefect In defect
            Dim p1 As CvPoint = item.Start, p2 As CvPoint = item.End
            Dim dist As Double = GetDistance(p1, p2)
            Dim mid As CvPoint2D64f = GetMidpoint(p1, p2)
            img.DrawLine(p1, p2, CvColor.White, 3)
            img.DrawCircle(item.DepthPoint, 10, CvColor.Green, -1)
            img.DrawLine(mid, item.DepthPoint, CvColor.White, 1)
            Form1.TextBox1.AppendText(Environment.NewLine & String.Format("No:{0} Depth:{1} Dist:{2}", count, item.Depth, dist))
            count += 1
        Next item
    End Sub
    ''' <summary>
    ''' 2点間の距離を得る
    ''' </summary>
    ''' <param name="p1"></param>
    ''' <param name="p2"></param>
    ''' <returns></returns>
    Private Function GetDistance(ByVal p1 As CvPoint, ByVal p2 As CvPoint) As Double
        Return Math.Sqrt(Math.Pow(p1.X - p2.X, 2) + Math.Pow(p1.Y - p2.Y, 2))
    End Function
    ''' <summary>
    ''' 2点の中点を得る
    ''' </summary>
    ''' <param name="p1"></param>
    ''' <param name="p2"></param>
    ''' <returns></returns>
    Private Function GetMidpoint(ByVal p1 As CvPoint, ByVal p2 As CvPoint) As CvPoint2D64f
        Return New CvPoint2D64f With {.X = (p1.X + p2.X) / 2.0, .Y = (p1.Y + p2.Y) / 2.0}
    End Function
End Module
' End Namespace
